src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,722 @@
+        else
+          None
+      }: _*),
+      colType = mt.colType.filter(colKeySet)._1,
+      globalType = TStruct(mt.globalType.required)
+    )
+  }
+
+  def minimal[T <: Type](base: T): T = {
+    val result = base match {
+      case ts: TStruct => TStruct(ts.required)
+      case ta: TArray => TArray(minimal(ta.elementType), ta.required)
+      case tagg: TAggregable => tagg.copy(elementType = minimal(tagg.elementType))
+      case t => t
+    }
+    result.asInstanceOf[T]
+  }
+
+  def unifyBaseType(base: BaseType, children: BaseType*): BaseType = {
+    base match {
+      case tt: TableType =>
+        val ttChildren = children.map(_.asInstanceOf[TableType])
+        tt.copy(
+          rowType = unify(tt.rowType, ttChildren.map(_.rowType): _*),
+          globalType = unify(tt.globalType, ttChildren.map(_.globalType): _*)
+        )
+      case mt: MatrixType =>
+        val mtChildren = children.map(_.asInstanceOf[MatrixType])
+        mt.copy(
+          globalType = unify(mt.globalType, mtChildren.map(_.globalType): _*),
+          rvRowType = unify(mt.rvRowType, mtChildren.map(_.rvRowType): _*),
+          colType = unify(mt.colType, mtChildren.map(_.colType): _*)
+        )
+      case t: Type =>
+        if (children.isEmpty)
+          return minimal(t)
+        t match {
+          case ts: TStruct =>
+            val subStructs = children.map(_.asInstanceOf[TStruct])
+            val subFields = ts.fields.map { f =>
+              f -> subStructs.flatMap(s => s.fieldOption(f.name))
+            }
+              .filter(_._2.nonEmpty)
+              .map { case (f, ss) => f.name -> unify(f.typ, ss.map(_.typ): _*) }
+            TStruct(ts.required, subFields: _*)
+          case tt: TTuple =>
+            val subTuples = children.map(_.asInstanceOf[TTuple])
+            TTuple(tt.required, tt.types.indices.map(i => unify(tt.types(i), subTuples.map(_.types(i)): _*)): _*)
+          case ta: TArray =>
+            TArray(unify(ta.elementType, children.map(_.asInstanceOf[TArray].elementType): _*), ta.required)
+          case tagg: TAggregable => tagg.copy(elementType =
+            unify(tagg.elementType, children.map(_.asInstanceOf[TAggregable].elementType): _*))
+          case _ =>
+            assert(children.forall(_.asInstanceOf[Type].isOfType(t)))
+            base
+        }
+    }
+  }
+
+  def unify[T <: BaseType](base: T, children: T*): T = unifyBaseType(base, children: _*).asInstanceOf[T]
+
+  def unifyEnvs(envs: Env[(Type, Type)]*): Env[(Type, Type)] = {
+    val lc = envs.lengthCompare(1)
+    if (lc < 0)
+      Env.empty[(Type, Type)]
+    else if (lc == 0)
+      envs.head
+    else {
+      val allKeys = envs.flatMap(_.m.keys).toSet
+      val bindings = allKeys.toArray.map { k =>
+        val envMatches = envs.flatMap(_.lookupOption(k))
+        assert(envMatches.nonEmpty)
+        val base = envMatches.head._1
+        k -> (base, unify(base, envMatches.map(_._2): _*))
```

Comment:
I'm concerned about this.  Isn't the first argument of unify special?  Base is selected randomly.  Do we need a reference environment with the full set of types?

Reply:
the _1 should be the same in every environment. we could assert this here, but it might be expensive.